File: python/copass-agent-router/src/copass_agent_router/connect_flow.py
"""End-to-end OAuth connect flow helper.

Mirror of the TS ``runConnectFlow``. Uses Python's ``http.server`` for
the short-lived success/error redirect listener and polls
``integrations.reconcile`` until the DataSource lands.

Server-side / CLI flows use this; webapp flows typically skip the
listener and handle redirects in the browser directly.
"""

from __future__ import annotations

import asyncio
import http.server
import logging
import socket
import threading
from dataclasses import dataclass
from typing import Any, Awaitable, Callable, Optional, Union

from copass_core import CopassClient

logger = logging.getLogger(__name__)


OnConnectUrl = Callable[[str], Union[None, Awaitable[None]]]


@dataclass(frozen=True)
class ConnectFlowResult:
    connection: dict
    session_id: str


class _RedirectListener:
    """One-shot HTTP listener on ``127.0.0.1:<random>`` capturing
    ``/oauth/success`` or ``/oauth/error``."""

    def __init__(self) -> None:
        self._outcome: Optional[str] = None
        self._event = threading.Event()
        self._server: Optional[http.server.HTTPServer] = None
        self._thread: Optional[threading.Thread] = None

    def start(self) -> tuple[str, str]:
        outer = self

        class _Handler(http.server.BaseHTTPRequestHandler):
            def do_GET(self) -> None:  # noqa: N802
                path = (self.path or "/").split("?")[0]
                if path == "/oauth/success":
                    outer._outcome = "success"
                    body = (
                        "<html><body style='font-family:sans-serif;padding:3rem'>"
                        "<h2>\u2713 Connection complete</h2>"
                        "<p>You can close this tab.</p></body></html>"
                    )
                elif path == "/oauth/error":
                    outer._outcome = "error"
                    body = (
                        "<html><body style='font-family:sans-serif;padding:3rem'>"
                        "<h2>\u2717 Connection failed</h2>"
                        "<p>You can close this tab and retry.</p></body></html>"
                    )
                else:
                    self.send_response(404)
                    self.end_headers()
                    return
                self.send_response(200)
                self.send_header("Content-Type", "text/html; charset=utf-8")
                self.end_headers()
                self.wfile.write(body.encode("utf-8"))
                outer._event.set()

            def log_message(self, *_args: Any) -> None:  # silence stdlib stdout
                return

        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        sock.bind(("127.0.0.1", 0))
        port = sock.getsockname()[1]
        sock.close()
        self._server = http.server.HTTPServer(("127.0.0.1", port), _Handler)
        self._thread = threading.Thread(target=self._server.serve_forever, daemon=True)
        self._thread.start()
        base = f"http://127.0.0.1:{port}"
        return f"{base}/oauth/success", f"{base}/oauth/error"

    async def wait(self, timeout_seconds: float) -> Optional[str]:
        """Return 'success' | 'error' | None (timeout)."""
        loop = asyncio.get_event_loop()
        await loop.run_in_executor(None, self._event.wait, timeout_seconds)
        return self._outcome

    def close(self) -> None:
        if self._server is not None:
            try:
                self._server.shutdown()
                self._server.server_close()
            except Exception:  # pragma: no cover
                pass


async def run_connect_flow(
    client: CopassClient,
    sandbox_id: str,
    *,
    app: str,
    on_connect_url: OnConnectUrl,
    scope: str = "user",
    project_id: Optional[str] = None,
    timeout_seconds: float = 300.0,
    success_uri: Optional[str] = None,
    error_uri: Optional[str] = None,
) -> ConnectFlowResult:
    """Run the Copass integrations connect flow end-to-end.

    1. Snapshots current connections for diff detection.
    2. Starts a localhost listener for the success redirect (unless
       caller supplied ``success_uri`` / ``error_uri``).
    3. Mints a Connect URL via :meth:`IntegrationsResource.connect`.
    4. Calls ``on_connect_url(url)`` — typically opens the browser.
    5. Polls ``integrations.reconcile`` every 2s until a new connection
       not in the pre-snapshot appears, OR the listener fires error, OR
       the timeout elapses.
    """
    # Snapshot
    try:
        existing = await client.integrations.list(sandbox_id, app=app)
        known_before = {c["source_id"] for c in existing.get("items", [])}
    except Exception:  # pragma: no cover — non-fatal
        known_before = set()

    listener: Optional[_RedirectListener] = None
    if success_uri is None or error_uri is None:
        listener = _RedirectListener()
        s, e = listener.start()
        success_uri = success_uri or s
        error_uri = error_uri or e

    try:
        resp = await client.integrations.connect(
            sandbox_id,
            app=app,
            scope=scope,
            success_redirect_uri=success_uri,
            error_redirect_uri=error_uri,
            project_id=project_id,
        )
        session_id = str(resp["session_id"])
        connect_url = str(resp["connect_url"])
        result = on_connect_url(connect_url)
        if asyncio.iscoroutine(result):
            await result

        # Poll reconcile until a new connection appears or the listener
        # signals 'error' or the timeout elapses.
        deadline = asyncio.get_event_loop().time() + timeout_seconds
        redirect_task: Optional[asyncio.Task[Optional[str]]] = (
            asyncio.create_task(listener.wait(timeout_seconds)) if listener else None
        )

        while asyncio.get_event_loop().time() < deadline:
            try:
                r = await client.integrations.reconcile(
                    sandbox_id, app=app, scope=scope
                )
                for c in r.get("connections", []):
                    if c["source_id"] not in known_before:
                        if redirect_task is not None:
                            redirect_task.cancel()
                        return ConnectFlowResult(
                            connection=dict(c), session_id=session_id
                        )
            except Exception:
                pass

            if redirect_task is not None and redirect_task.done():
                outcome = redirect_task.result()
                if outcome == "error":
                    raise RuntimeError(
                        "User denied the authorization or provider returned an error."
                    )

            await asyncio.sleep(2.0)

        raise TimeoutError(
            f"Timed out after {int(timeout_seconds)}s — the connection may still "
            "land. Check `client.integrations.list(sandbox_id)` or run reconcile."
        )
    finally:
        if listener is not None:
            listener.close()


__all__ = ["run_connect_flow", "ConnectFlowResult", "OnConnectUrl"]

File: typescript/packages/agent-router/src/connect-flow.ts
/**
 * Provider-OAuth flow helper — abstracts the "mint Connect URL, redirect
 * browser, poll for webhook-driven DataSource" dance into one call.
 *
 * Node-only: uses `node:http` for the short-lived success/error redirect
 * listener. Browsers should call `integrations.connect()` on `@copass/core`
 * directly and handle the redirect UX themselves (typically via a popup
 * + BroadcastChannel), so this module stays out of the browser bundle.
 */

import http, { type IncomingMessage, type ServerResponse } from 'node:http';
import type { AddressInfo } from 'node:net';
import type {
  ConnectionItem,
  CopassClient,
  IntegrationScope,
} from '@copass/core';

export interface ConnectFlowOptions {
  /** App slug (`slack_v2`, `github`, `notion`, …). Must be supported. */
  app: string;
  /** Default `'user'`. */
  scope?: IntegrationScope;
  /** Override project. */
  projectId?: string;
  /** Called with the hosted Connect URL — open a browser / print / webview. */
  onConnectUrl: (url: string) => void | Promise<void>;
  /** Max seconds waiting for the user to complete OAuth + webhook/reconcile. */
  timeoutSeconds?: number;
  /** Override success redirect (default: local listener). */
  successUri?: string;
  /** Override error redirect (default: local listener). */
  errorUri?: string;
  /** Called on each poll tick with the current reconcile result. */
  onPoll?: (info: { createdThisTick: number; elapsed: number }) => void;
}

export interface ConnectFlowResult {
  connection: ConnectionItem;
  sessionId: string;
}

/** Stand up a one-shot localhost HTTP listener on a random port. */
async function startRedirectListener(
  timeoutMs: number,
): Promise<{
  successUri: string;
  errorUri: string;
  waitForRedirect: () => Promise<'success' | 'error'>;
  close: () => void;
}> {
  let resolveRedirect!: (v: 'success' | 'error') => void;
  let rejectRedirect!: (e: Error) => void;
  const redirectPromise = new Promise<'success' | 'error'>((resolve, reject) => {
    resolveRedirect = resolve;
    rejectRedirect = reject;
  });

  const server = http.createServer((req: IncomingMessage, res: ServerResponse) => {
    const url = new URL(req.url ?? '/', 'http://localhost');
    if (url.pathname === '/oauth/success' || url.pathname === '/oauth/error') {
      const outcome = url.pathname === '/oauth/success' ? 'success' : 'error';
      const body =
        outcome === 'success'
          ? '<html><body style="font-family:sans-serif;padding:3rem"><h2>\u2713 Connection complete</h2><p>You can close this tab.</p></body></html>'
          : '<html><body style="font-family:sans-serif;padding:3rem"><h2>\u2717 Connection failed</h2><p>You can close this tab and retry.</p></body></html>';
      res.writeHead(200, { 'Content-Type': 'text/html; charset=utf-8' });
      res.end(body);
      resolveRedirect(outcome);
    } else {
      res.writeHead(404);
      res.end();
    }
  });

  await new Promise<void>((resolve) => server.listen(0, '127.0.0.1', resolve));
  const port = (server.address() as AddressInfo).port;
  const base = `http://127.0.0.1:${port}`;

  const timeout = setTimeout(() => {
    rejectRedirect(
      new Error(
        `Timed out after ${Math.round(timeoutMs / 1000)}s waiting for OAuth redirect.`,
      ),
    );
  }, timeoutMs);

  const close = () => {
    clearTimeout(timeout);
    try {
      server.close();
    } catch {
      /* ignore */
    }
  };

  return {
    successUri: `${base}/oauth/success`,
    errorUri: `${base}/oauth/error`,
    waitForRedirect: async () => {
      try {
        return await redirectPromise;
      } finally {
        close();
      }
    },
    close,
  };
}

/** Run the Copass integration connect flow end-to-end. Returns the new
 * connection once it lands, or throws on timeout / user denial. */
export async function runConnectFlow(
  client: CopassClient,
  sandboxId: string,
  options: ConnectFlowOptions,
): Promise<ConnectFlowResult> {
  const scope: IntegrationScope = options.scope ?? 'user';
  const app = options.app;
  const timeoutMs = Math.max(30, options.timeoutSeconds ?? 300) * 1000;

  // Snapshot existing connections so we can detect the new one.
  let knownBefore = new Set<string>();
  try {
    const existing = await client.integrations.list(sandboxId, { app });
    knownBefore = new Set(existing.items.map((c) => c.source_id));
  } catch {
    /* non-fatal — reconcile will still work */
  }

  const listener = await startRedirectListener(timeoutMs);
  const successUri = options.successUri ?? listener.successUri;
  const errorUri = options.errorUri ?? listener.errorUri;

  let sessionId = '';
  try {
    const resp = await client.integrations.connect(sandboxId, app, {
      scope,
      success_redirect_uri: successUri,
      error_redirect_uri: errorUri,
      project_id: options.projectId,
    });
    sessionId = resp.session_id;
    await options.onConnectUrl(resp.connect_url);
  } catch (err) {
    listener.close();
    throw err;
  }

  // Race the browser redirect against the reconcile poll — the webhook
  // can race us too, and that's fine; reconcile() is idempotent.
  const start = Date.now();
  const redirectPromise = listener.waitForRedirect().catch(() => 'timeout' as const);

  while (Date.now() - start < timeoutMs) {
    try {
      const r = await client.integrations.reconcile(sandboxId, { app, scope });
      for (const c of r.connections) {
        if (!knownBefore.has(c.source_id)) {
          listener.close();
          return { connection: c, sessionId };
        }
      }
      options.onPoll?.({ createdThisTick: r.created_count, elapsed: Date.now() - start });
    } catch {
      /* ignore transient errors */
    }

    // Yield briefly. If the redirect fires mid-sleep, we short-circuit
    // on the next loop iteration.
    const sleep = new Promise<'tick'>((r) => setTimeout(() => r('tick'), 2000));
    const winner = await Promise.race([
      redirectPromise.then((v) => (v === 'error' ? 'error' : 'redirect')),
      sleep,
    ]);
    if (winner === 'error') {
      listener.close();
      throw new Error('User denied the authorization or provider returned an error.');
    }
    // 'redirect' or 'tick' — either way, loop back and reconcile again.
  }

  listener.close();
  throw new Error(
    `Timed out after ${Math.round(timeoutMs / 1000)}s. The connection may still land — ` +
      'check `client.integrations.list()` in a moment or run reconcile manually.',
  );
}
